File: dynamic-programming/binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findMaxSum(self, root:Optional[TreeNode], pathSum:List) ->int:
        if not root:
            return 0
        
        lSum = max(0, self.findMaxSum(root.left, pathSum))
        rSum = max(0, self.findMaxSum(root.right, pathSum))
        pathSum[0] = max(pathSum[0], lSum+rSum+root.val)
        
        return max(lSum, rSum)+root.val
    
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        pathSum = [float('-inf')]
        self.findMaxSum(root, pathSum)
        return pathSum[0]

        
        

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.go
func max(x int, y int) int {
    if x > y {
        return x
    }
    return y
}

func calc(root *TreeNode, sum *int) int {
    if root == nil {
        return 0
    }

    left := max(0, calc(root.Left, sum))
    right := max(0, calc(root.Right, sum))

    *sum = max(*sum, left+right+root.Val)

    return root.Val + max(left, right)
}

func maxPathSum(root *TreeNode) int {
    sum := root.Val
    calc(root, &sum)
    return sum
}
